grpc-proto/proto/schema.proto

```diff
@@ -0,0 +1,211 @@
+  POLYGON = 0x10;
+  SMALLINT = 0x11;
+  TEXT = 0x12;
+  TIME = 0x13;
+  TIMESTAMP = 0x14;
+  TIMEUUID = 0x15;
+  TINYINT = 0x16;
+  TUPLE = 0x17;
+  UDT = 0x18;
+  UUID = 0x19;
+  VARCHAR = 0x1A;
+  VARINT = 0x1B;
+}
+
+// USER DEFINED TYPES
+
+message CqlUdtField {
+  string name = 1;
+
+  CqlType type = 2;
+
+  // only used for UDTs (UDTs can reference other UDTs, but they must be frozen)
+  string cqlTypeName = 3;
+
+  bool frozen = 4;
+}
+
+message CqlUserDefinedType {
+  string name = 1;
+
+  repeated CqlUdtField fields = 2;
+
+}
+
+// COLUMNS
+
+enum CqlColumnOrderBy {
+  ASC = 0x00;
+  DESC = 0x01;
+}
+
+message CqlComplexColumnType {
+  CqlType type = 1;
+
+  // only used if type=UDT
+  string udtTypeName = 2;
+
+  bool frozen = 3;
+}
+
+message CqlColumn {
+
+  string name = 1;
+
+  CqlType type = 2;
+
+  // only used if type=UDT
+  string udtTypeName = 3;
+
+  // only used if type is one of {SET, LIST, MAP, TUPLE}
+  // for SET and LIST, this will have one element
+  //   example: direct_reports set<frozen <fullname>> -> CqlComplexColumnType(UDT, fullname, true)
+  //   example: prime_numbers list<int> -> CqlComplexColumnType(INT, null, false)
+  // for MAP, this will have two elements, the key type, followed by the value type
+  //   example: addresses map<string, frozen<address>> -> CqlComplexColumnType(STRING, null, false),
+  //     CqlComplexColumnType(UDT, address, true)
+  // for TUPLE, this will have as many elements as are referenced
+  //   example: values <tuple<int, text, float>> -> CqlComplexColumnType(INT, null, false),
+  //     CqlComplexColumnType(STRING, null, false), CqlComplexColumnType(FLOAT, null, false)
+  repeated CqlComplexColumnType complexColumnTypes = 4;
+
+  bool isPartitionKey = 5;
+
+  // only used for clustering columns
+  uint32 clusteringColumnOrder = 6;
+
+  bool isFrozen = 7;
+
+  bool isStatic = 8;
+}
+
+// TABLES
+
+// TODO ok to make it this simplistic?
+message CqlTableOption {
+
+  string name = 1;
+  string value = 2;
+}
+
+message CqlTable {
+
+  string name = 1;
+
+  repeated CqlColumn columns = 2;
+
+  repeated CqlTableOption tableOptions = 3;
+}
+
+// INDEXES
+
+message CqlIndex {
+  // TODO
+}
+
+// MATERIALIZED VIEWS
+
+message CqlMaterializedView {
+  // TODO
+}
+
+// KEYSPACES
+
+message CqlDatacenterReplication {
+  string datacenterName = 1;
+
+  uint32 replicationFactor = 2;
+}
+
+message CqlKeyspace {
+
+  string name = 1;
+
+  string replicationStrategy = 2;
+
+  repeated CqlDatacenterReplication replicationFactors = 3;
+
+  bool durableWrites = 4;
+
+}
+
+// DESCRIBE KEYSPACE
+
+message DescribeQuery {
+  string keyspaceName = 1;
+}
+
+message CqlKeyspaceDescribe {
+
+  CqlKeyspace cqlKeyspace = 1;
+
+  repeated CqlUserDefinedType types = 2;
+
+  repeated CqlTable tables = 3;
+
+  repeated CqlIndex indexes = 4;
+
+  repeated CqlMaterializedView materializedViews = 5;
+}
+
+// SCHEMA CREATION
+
+message CqlSchemaOperation {
+  oneof schema {
+
+    CqlTable table = 1;
```

Comment:
This might be missing a create keyspace operation.

Reply:
will add